### PR TITLE
Entities bug-fixes in place management

### PR DESCRIPTION
**Pull request by benbosman**
We've noticed a few use cases where the place of a new relationship is not defined correctly. These problems have been solved and JUnit tests have been created to automatically verify the functionality.

The previously existing tests didn't cause any problems because:
* All tests added the relationship as the last element
* There were no tests to change the place of a relationship
* The item.update was always called somewhere implicitly to fix the places

All of the previous tests still succeed, and the additional tests succeed as well.
Inline comments have been included to explain the more complicated reasons for the approach in determining the place
The JavaDocs have been updated as well

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java [hunk 1]
```diff
@@ -778,8 +778,8 @@ private void addRelationship(Context c, Item item, String typeName, String value
         }
 
         // Create the relationship
-        int leftPlace = relationshipService.findLeftPlaceByLeftItem(c, leftItem) + 1;
-        int rightPlace = relationshipService.findRightPlaceByRightItem(c, rightItem) + 1;
+        int leftPlace = relationshipService.findNextLeftPlaceByLeftItem(c, leftItem);
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: The first time I executed an CSV import, I got an error. On the second try, without any changes, the record was registered. But I noticed a duplication of isAuthorOfPublication on the Metadatafield with the relationship.

`metadata_field_id: 134 (relation.isAuthorOfPublication)`
2019-10-29 09_52_48-Valentina Studio

Should a specific relation.isAuthorOfPublication should be present in this table?

Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/app/bulkedit/MetadataImport.java]: Hi @paulo-graca,

I was able to reproduce the problem, but also on master. It seems you've found a bug in one of the CSV branches already merged in master.
Can you move this to a separate ticket so it's not lost?

I was able to reproduce this with the CSV https://www.dropbox.com/s/vvugb1tuknjhp9a/test-records-2.csv?dl=1 which created relation.isAuthorOfPublication with value 0ffbee3f-e7ea-42bc-92fe-2fbef1a52c0f twice using master

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java [hunk 2]
```diff
@@ -94,18 +98,32 @@ public Relationship create(Context context, Relationship relationship) throws SQ
...
         throws SQLException, AuthorizeException {
         Item leftItem = relationship.getLeftItem();
+        // Max value is used to ensure that these will get added to the back of the list and thus receive the highest
+        // (last) place as it's set to a -1 for creation
+        if (relationship.getLeftPlace() == -1) {
+            relationship.setLeftPlace(Integer.MAX_VALUE);
+        }
+        Item rightItem = relationship.getRightItem();
+        if (relationship.getRightPlace() == -1) {
+            relationship.setRightPlace(Integer.MAX_VALUE);
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: This method's logic  is increasingly confusing.  Can you better explain why we have to temporarily set the place to `MAX_VALUE`?  Why not just set it to the length of `leftRelationships` or `rightRelationships` on the next lines?

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java [hunk 3]
```diff
@@ -94,18 +98,32 @@ public Relationship create(Context context, Relationship relationship) throws SQ
...
         List<Relationship> rightRelationships = findByItemAndRelationshipType(context,
                                                                               rightItem,
                                                                               relationship.getRelationshipType(),
                                                                               false);
 
+        if (leftRelationships.contains(relationship)) {
+            leftRelationships.remove(relationship);
+        }
+        if (rightRelationships.contains(relationship)) {
+            rightRelationships.remove(relationship);
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: These if statements should be preceded by a comment to describe why we are removing the relationship from the list.  I don't understand the logic here.  Could we add some Unit Tests to prove out this is working as expected?

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+     */
+    @After
+    @Override
+    public void destroy() {
+        context.abort();
+        super.destroy();
+    }
+
+    @Test
+    public void addMetadataAndRelationshipTest() throws Exception {
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Either break up this method into separate tests, or please add inline comments.  This method is hard to follow as you create new content midway through and then run more tests.  It's difficult for me to see what exactly you are testing for in each test without understanding all 50 lines of this method.

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        assertMetadataValue(authorQualifier, contributorElement, dcSchema, "test, three", null, 3, list.get(3));
+        assertMetadataValue(authorQualifier, contributorElement, dcSchema, "test, four", null, 4, list.get(4));
+        assertMetadataValue(authorQualifier, contributorElement, dcSchema, "familyNameTwo, firstNameTwo",
+                            "virtual::" + relationshipTwo.getID(), 5, list.get(5));
+        assertThat(relationshipTwo.getLeftPlace(), equalTo(5));
+
+    }
+
+    @Test
+    public void AddMetadataAndRelationshipWithSpecificPlaceTest() throws Exception {
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Again, method is too long, and you add new content at several stages.  Please break this up into several smaller tests (with more descriptive names)...or add descriptive inline comments

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        assertMetadataValue(authorQualifier, contributorElement, dcSchema, "test, four", null, 4, list.get(4));
+        assertMetadataValue(authorQualifier, contributorElement, dcSchema, "familyNameTwo, firstNameTwo",
+                            "virtual::" + relationshipTwo.getID(), 5, list.get(5));
+        assertThat(relationshipTwo.getLeftPlace(), equalTo(5));
+
+    }
+
+
+    @Test
+    public void AddAndRemoveMetadataAndRelationshipsTest() throws Exception {
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Same as above.  Too large and no comments

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+        metadataValueService.delete(context, metadataValueToRemove);
+
+        context.restoreAuthSystemState();
+
+        List<MetadataValue> list = itemService
+            .getMetadata(item, dcSchema, contributorElement, authorQualifier, Item.ANY);
+        assertThat(list.size(), equalTo(2));
+
+        // Check that these places are still intact after the deletion as the place doesn't get updated until an
+        // item.update has been called
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: This is a better inline comment.  But, where do we call `item.update` below to ensure the place *does* get updated after the deletion?

**Review by tdonohue — COMMENTED** [review]
@benbosman : I gave this a quick review today. I'm confused by several of the code changes here (see below).  As the new test methods are very large an uncommented, it's difficult to follow them as well.  I'd appreciate either inline comments in large test methods, or find a way to break them up into smaller these methods (where plausible).

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java (hunk 2 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: The place -1 implies it's not set and should be appended at the end.
The length of the arrays is also not necessarily the correct place, e.g. if there are plain text authors present

A comment has been included at https://github.com/DSpace/DSpace/pull/2551/files/e56e71a90338000666a2ce012c3849fc56d0c639#diff-2e89fe18482e7ac43a1a4e901a8105b7R104 to explain this

File: dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java [hunk 8]
```diff
@@ -86,9 +86,9 @@ public int findRightPlaceByRightItem(Context context, Item item) throws SQLExcep
         List<Relationship> list = list(context, criteriaQuery, false, Relationship.class, -1, -1);
         list.sort((o1, o2) -> o2.getLeftPlace() - o1.getLeftPlace());
         if (!list.isEmpty()) {
-            return list.get(0).getLeftPlace();
+            return list.get(0).getLeftPlace() + 1;
```
[the remark this answers is not in the document]
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java]: we'll check this out

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java (hunk 7 above)
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: it happens implicitly in https://github.com/DSpace/DSpace/pull/2551/files/e56e71a90338000666a2ce012c3849fc56d0c639#diff-56b0380ff7b66daeff6012e659df5863R246 but it can be duplicated explicitly as well if that makes it easier to understand

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java (hunk 3 above)
Comment (Raf-atmire) [reply dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: These relationships here aren't removed from the actual list, they're only removed in this temporary list so that we can perform a proper place calculation further on in this method. I do agree that it was a bit unclear, so I'll add a comment stating this. This is particularly useful incase this method is called for a delete use case, so that we can update the relationships' places properly without the relationship given interfering with the place calculation

File: dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java]: This has been solved.
The `testGetNextRightPlace` method has been created to verify this works correctly

File: dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java (hunk 3 above)
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/content/RelationshipServiceImpl.java]: This has been clarified using inline comments

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java (hunk 4 above)
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Inline comments have been included

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java (hunk 5 above)
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Inline comments have been included

File: dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java (hunk 6 above)
Comment (benbosman) [reply dspace-api/src/test/java/org/dspace/content/RelationshipServiceImplPlaceTest.java]: Inline comments have been included

**Review by paulo-graca — APPROVED** [review]
The changes look ok to me. And didn't found any issues (besides the CSV import) within my tests. My thanks to @benbosman and @Raf-atmire for this contribution!

File: dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceTest.java [hunk 9]
```diff
@@ -138,5 +138,41 @@ public void testGetRelationshipMetadata() throws Exception {
         assertThat(list.get(1).getMetadataField().getElement(), equalTo("isAuthorOfPublication"));
         assertThat(list.get(1).getAuthority(), equalTo("virtual::" + relationship.getID()));
 
+    }
+
+    @Test
+    public void testGetNextRightPlace() throws Exception {
```
Comment (tdonohue) [inline dspace-api/src/test/java/org/dspace/content/RelationshipMetadataServiceTest.java]: Could we add a corresponding test for `findNextLeftPlaceByLeftItem`?  I just noticed it seems to be missing.

**Review by tdonohue — COMMENTED** [review]
@benbosman : This is looking good overall now.  Just one (very minor) missing test...see below.  Once that's ready, I think this can be merged.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @benbosman and @Raf-atmire.  Looks good to me!
